draft-ietf-quic-http.md

```diff
@@ -180,103 +180,232 @@ frames, exposing the data contained within as a reliable byte stream to the
+({{stream-request}}). A server sends an HTTP response on a
+response stream ({{stream-response}}).  The response stream header includes the
+stream ID of the request stream.
 
 An HTTP message (request or response) consists of:
 
-1. one header block (see {{frame-headers}}) on the control stream containing the
-   message headers (see {{!RFC7230}}, Section 3.2),
+1. one header block (see {{frame-headers}}) on the request or response stream
+   containing the message headers (see {{!RFC7230}}, Section 3.2),
 
-2. the payload body (see {{!RFC7230}}, Section 3.3), sent on the data stream,
+2. optionally, the payload body (see {{!RFC7230}}, Section 3.3), which is sent
+   on the data stream if the request or response stream includes a HAS_BODY
+   frame after the initial header block (see {{frame-has-body}}),
 
-3. optionally, one header block on the control stream containing the
+3. optionally, one header block on the request or response stream containing the
    trailer-part, if present (see {{!RFC7230}}, Section 4.1.2).
 
 In addition, prior to sending the message header block indicated above, a
-response may contain zero or more header blocks on the control stream containing
-the message headers of informational (1xx) HTTP responses (see {{!RFC7230}},
-Section 3.2 and {{!RFC7231}}, Section 6.2).
-
-The data stream MUST be half-closed immediately after the transfer of the body.
-If the message does not contain a body, the corresponding data stream MUST still
-be half-closed without transferring any data. The "chunked" transfer encoding
-defined in Section 4.1 of {{!RFC7230}} MUST NOT be used.
-
-Trailing header fields are carried in an additional header block on the message
-control stream. Such a header block is a sequence of HEADERS frames with End
-Header Block set on the last frame. Senders MUST send only one header block in
-the trailers section; receivers MUST decode any subsequent header blocks in
-order to maintain HPACK decoder state, but the resulting output MUST be
-discarded.
-
-An HTTP request/response exchange fully consumes a pair of streams. After
-sending a request, a client closes the streams for sending; after sending a
-response, the server closes its streams for sending and the QUIC streams are
-fully closed.
+response may contain zero or more header blocks on the response stream
+containing the message headers of informational (1xx) HTTP responses (see
+{{!RFC7230}}, Section 3.2 and {{!RFC7231}}, Section 6.2).
+
+A message that contains a payload body includes a HAS_BODY frame immediately
+after the first header block.  This indicates that another stream will include
+the payload body.  A data stream is opened to carry the payload body.  The data
+stream MUST be closed immediately after sending the body.  If the message does
+not contain a body, a data stream MUST NOT be opened and the HAS_BODY message
+MUST NOT be sent.  The "chunked" transfer encoding defined in Section 4.1 of
+{{!RFC7230}} MUST NOT be used.
```

Comment:
I wanted to be able to create the body at any time, without knowing what stream might be used.  It's a minor thing, but it also keeps all pointers as back references.